app/src/main/java/org/wikipedia/page/edit_history/EditHistoryListActivity.kt

```diff
@@ -0,0 +1,166 @@
+        var listItems = mutableListOf<Any>()
+
+        init {
+            setUpList()
+        }
+
+        override fun getItemViewType(position: Int): Int {
+            return if (listItems[position] is Revision) VIEW_TYPE_ITEM else VIEW_TYPE_HEADER
+        }
+
+        override fun getItemCount(): Int {
+            return listItems.size
+        }
+
+        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
+            val inflater = LayoutInflater.from(this@EditHistoryListActivity)
+            return if (viewType == VIEW_TYPE_HEADER) {
+                HeaderViewHolder(inflater.inflate(layout.edit_history_section_header, parent, false))
+            } else {
+                EditHistoryListItemHolder(inflater.inflate(layout.item_edit_history, parent, false))
+            }
+        }
+
+        override fun onBindViewHolder(holder: ViewHolder, pos: Int) {
+            if (holder is HeaderViewHolder) {
+                holder.bindItem(listItems[pos] as String)
+            } else if (holder is EditHistoryListItemHolder) {
+                holder.bindItem(getOldRevision(pos), listItems[pos] as Revision)
+                holder.itemView.setOnClickListener(this)
+            }
+            holder.itemView.tag = pos
+        }
+
+        private fun getOldRevision(position: Int): Revision? {
+            return if (position < listItems.size - 1) {
+                if (listItems[position + 1] is Revision) listItems[position + 1] as Revision
+                else listItems[position + 2] as Revision
+            } else null
+        }
+
+        fun setUpList() {
+            editHistoryList.forEach {
+                val dateStr =
+                    DateUtil.getMonthOnlyDateString(DateUtil.iso8601DateParse(it.timeStamp))
+                if (!listItems.contains(dateStr)) {
+                    listItems.add(dateStr)
+                }
+                listItems.add(it)
+            }
+        }
+
+        override fun onClick(v: View) {
+            val item = listItems[v.tag as Int]
+            if (item is Revision) {
+                startActivity(ArticleEditDetailsActivity.newIntent(this@EditHistoryListActivity, pageTitle.prefixedText, item.revId, pageTitle.wikiSite.languageCode))
+            }
+        }
+    }
+
+    private inner class HeaderViewHolder constructor(itemView: View) :
+        ViewHolder(itemView) {
+        fun bindItem(listItem: String) {
+            itemView.findViewById<TextView>(id.section_header_text).text = listItem
+        }
+    }
+
+    private inner class EditHistoryListItemHolder constructor(itemView: View) :
+        ViewHolder(itemView) {
+        fun bindItem(oldRevision: Revision?, listItem: Revision) {
+            CoroutineScope(Dispatchers.IO).launch(CoroutineExceptionHandler { _, msg -> run { L.e(msg) } }) {
+                val editDetails: EditDetails = viewModel.fetchEditDetails(pageTitle.wikiSite.languageCode, oldRevision?.revId ?: 0, listItem.revId)
```

Comment:
I'm not seeing the necessity of making a separate call to fetch the details of each specific revision. Don't we already have all this data in the list of revisions that you fetch initially?

Reply:
Correct @dbrant. I will fetch it on need-to basis, but yes talking to Robin about how best to handle the ui without having to make another call for each row, especially since the click takes them to the full diff details view. Your suggestion on the task is a good middle ground.

Reply:
@dbrant I am still making the api call for the diffSize, wondering if that should be removed as well, and just show in bytes, which is part of the first call anyway?

Reply:
@sharvaniharan Absolutely, let's use the `size` field that we get for free from the initial call.